remote-vfs.c

```diff
@@ -0,0 +1,1889 @@
+
+	end_url_with_slash(&rest_url, url_base);
+	strbuf_addstr(&rest_url, url_component);
+
+	headers = curl_slist_append(headers,
+				    "X-TFS-FedAuthRedirect: Suppress");
+
+	slot = get_active_slot();
+
+	curl_easy_setopt(slot->curl, CURLOPT_URL, rest_url.buf);
+	curl_easy_setopt(slot->curl, CURLOPT_HTTPHEADER, headers);
+	curl_easy_setopt(slot->curl, CURLOPT_WRITEFUNCTION, fwrite);
+	curl_easy_setopt(slot->curl, CURLOPT_WRITEDATA, (void*)tempfile->fp);
+
+	if (creds && creds->username) {
+		/*
+		 * Force CURL to respect the username/password we provide by
+		 * turning off the AUTH-ANY negotiation stuff.
+		 *
+		 * (AUTH-ANY forces a 401 regardless of what we send).
+		 */
+		curl_easy_setopt(slot->curl, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
+		curl_easy_setopt(slot->curl, CURLOPT_USERNAME, creds->username);
+		curl_easy_setopt(slot->curl, CURLOPT_PASSWORD, creds->password);
+	} else {
+		/*
+		 * Turn on the AUTH-ANY negotiation.  This only works with the
+		 * main Git server (because the cache-server doesn't handle 401s).
+		 */
+		curl_easy_setopt(slot->curl, CURLOPT_HTTPAUTH, CURLAUTH_ANY);
+	}
+
+	ret = run_one_slot(slot, &results);
+	curl_easy_getinfo(slot->curl, CURLINFO_RESPONSE_CODE,
+			  http_response_code);
+
+//	trace2_printf("GET: [ret %d] [http %ld] [curl %d '%s'] [res %ld]",
+//		      ret, results.http_connectcode, results.curl_result,
+//		      curl_easy_strerror(results.curl_result),
+//		      *http_response_code);
+
+	strbuf_setlen(error_message, 0);
+	strbuf_addstr(error_message, curl_easy_strerror(results.curl_result));
+
+
+	curl_slist_free_all(headers);
+
+	return ret;
+}
+
+static int do_get__into_file__to_main(const char *url_component,
+				      struct tempfile *tempfile,
+				      struct strbuf *error_message)
+{
+	int ret = HTTP_OK;
+	const struct credential *p_creds = NULL;
+	long http_response_code;
+
+	lookup_main_creds();
+
+	if (rv_data.main_creds.username && *rv_data.main_creds.username)
+		p_creds = &rv_data.main_creds;
+
+	strbuf_setlen(error_message, 0);
+	trace2_region_enter("remote-vfs", "GET/object/main1", NULL);
+	ret = do_get__into_file(rv_data.buf_main_url.buf, url_component,
+				p_creds, tempfile,
+				&http_response_code, error_message);
+	trace2_region_leave("remote-vfs", "GET/object/main1", NULL);
+	if (ret == HTTP_OK)
+		return HTTP_OK;
+
+	if (ret != HTTP_REAUTH)
+		return ret;
+
+	refresh_main_creds();
+
+	/*
+	 * Since we are automatically retrying, truncate the error
+	 * message from the first attempt.  The caller should only
+	 * see the error message from the final attempt.
+	 */
+	strbuf_setlen(error_message, 0);
+
+	// TODO ftruncate tempfile ??
+
+	p_creds = &rv_data.main_creds;
+	trace2_region_enter("remote-vfs", "GET/object/main2", NULL);
+	ret = do_get__into_file(rv_data.buf_main_url.buf, url_component,
+				p_creds, tempfile,
+				&http_response_code, error_message);
+	trace2_region_leave("remote-vfs", "GET/object/main2", NULL);
+	if (ret == HTTP_OK)
+		return HTTP_OK;
```

Comment:
Again, deleting these two lines will make the code easier to read, without actually changing the behavior.

Reply:
again, this is for debugging and breakpoints.  the compiler should be able to flatten it out.

Reply:
Sure. The compiler can optimize this. I, however, cannot optimize this, when reviewing, and my eyes and my mind tires from such repetition...